File: diameter.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class BinaryTree:
    def __init__(self):
        self.diameter = 0

    # Method to calculate the height of the tree
    def calculate_height(self, node):
        if not node:
            return 0

        left_height = self.calculate_height(node.left)
        right_height = self.calculate_height(node.right)

        # Update the diameter
        self.diameter = max(self.diameter, left_height + right_height)

        # Return the height of the current node
        return max(left_height, right_height) + 1

# Example usage
if __name__ == "__main__":
    # Create the binary tree
    root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(3)
    root.left.left = TreeNode(4)
    root.left.right = TreeNode(5)

    # Find and print the diameter of the binary tree
    tree = BinaryTree()
    tree.calculate_height(root)  # Calculates height and updates diameter
    print("Diameter of the binary tree:", tree.diameter)

File: diameter.java
class TreeNode {
    int val;
    TreeNode left, right;

    TreeNode(int val) {
        this.val = val;
        left = right = null;
    }
}

class BinaryTree {
    private int diameter = 0;

    // Function to calculate the diameter of the binary tree
    public int diameterOfBinaryTree(TreeNode root) {
        calculateHeight(root);
        return diameter;
    }

    // Helper function to calculate the height of the tree
    private int calculateHeight(TreeNode node) {
        if (node == null) {
            return 0;
        }

        int leftHeight = calculateHeight(node.left);
        int rightHeight = calculateHeight(node.right);

        // Update diameter at each node
        diameter = Math.max(diameter, leftHeight + rightHeight);

        // Return height of the current node
        return Math.max(leftHeight, rightHeight) + 1;
    }
}

public class Main {
    public static void main(String[] args) {
        BinaryTree tree = new BinaryTree();

        // Create the binary tree
        TreeNode root = new TreeNode(1);
        root.left = new TreeNode(2);
        root.right = new TreeNode(3);
        root.left.left = new TreeNode(4);
        root.left.right = new TreeNode(5);

        // Find and print the diameter of the binary tree
        System.out.println("Diameter of the binary tree: " + tree.diameterOfBinaryTree(root));
    }
}
